File: app14_LUA_UNITY/frontend/Assets/StreamingAssets/lua/i18n.lua
-- D-05: a hand-written i18n store, mirroring app13_ruby_FastApi's i18n.js —
-- loaded independently from the backend's own i18n error-message table
-- (PLAN.md D-05). PL default, instant switch, no scene reload needed.
-- NFR-06: every key below MUST exist in both locales — enforced by
-- Tests/Lua/spec/i18n_spec.lua diffing the two tables' key sets.

local STRINGS = {
  pl = {
    ["nav.threats"] = "Zagrożenia",
    ["nav.cards"] = "Karty",
    ["nav.game"] = "Gra",
    ["nav.settings"] = "Ustawienia",
    ["login.title"] = "Logowanie",
    ["login.username"] = "Nazwa użytkownika",
    ["login.password"] = "Hasło",
    ["login.submit"] = "Zaloguj",
    ["login.error"] = "Nieprawidłowa nazwa użytkownika lub hasło.",
    ["threats.title"] = "Zagrożenia",
    ["threats.filter.severity"] = "Poziom istotności",
    ["threats.filter.all"] = "Wszystkie",
    ["threats.empty"] = "Brak wyników dla podanych filtrów.",
    ["detail.overview"] = "Przegląd",
    ["detail.attackVectors"] = "Wektory ataku",
    ["detail.mitigations"] = "Mitigacje",
    ["codeSample.attackDemo.label"] = "ATTACK DEMO — kod podatny, nie używać w produkcji",
    ["codeSample.attackDemo.reveal"] = "Pokaż kod (PODATNY)",
    ["codeSample.attackDemo.confirmTitle"] = "Ten kod celowo demonstruje podatność.",
    ["codeSample.attackDemo.confirm"] = "Rozumiem",
    ["codeSample.attackDemo.cancel"] = "Anuluj",
    ["game.mode.regular"] = "Tryb Regularny",
    ["game.mode.shiftLeft"] = "Tryb Shift Left",
    ["game.mode.workshop"] = "Warsztat Modelowania Zagrożeń",
    ["game.reputation"] = "Reputacja",
    ["game.turn"] = "Tura",
    ["game.victory"] = "Zwycięstwo!",
    ["game.defeat"] = "Porażka."
  },
  en = {
    ["nav.threats"] = "Threats",
    ["nav.cards"] = "Cards",
    ["nav.game"] = "Game",
    ["nav.settings"] = "Settings",
    ["login.title"] = "Login",
    ["login.username"] = "Username",
    ["login.password"] = "Password",
    ["login.submit"] = "Log in",
    ["login.error"] = "Invalid username or password.",
    ["threats.title"] = "Threats",
    ["threats.filter.severity"] = "Severity",
    ["threats.filter.all"] = "All",
    ["threats.empty"] = "No results for the current filters.",
    ["detail.overview"] = "Overview",
    ["detail.attackVectors"] = "Attack Vectors",
    ["detail.mitigations"] = "Mitigations",
    ["codeSample.attackDemo.label"] = "ATTACK DEMO — vulnerable code, do not use in production",
    ["codeSample.attackDemo.reveal"] = "Show code (VULNERABLE)",
    ["codeSample.attackDemo.confirmTitle"] = "This code deliberately demonstrates a vulnerability.",
    ["codeSample.attackDemo.confirm"] = "I understand",
    ["codeSample.attackDemo.cancel"] = "Cancel",
    ["game.mode.regular"] = "Regular Mode",
    ["game.mode.shiftLeft"] = "Shift Left Mode",
    ["game.mode.workshop"] = "Threat Modeling Workshop",
    ["game.reputation"] = "Reputation",
    ["game.turn"] = "Turn",
    ["game.victory"] = "Victory!",
    ["game.defeat"] = "Defeat."
  }
}

local i18n = {}
local current_locale = "pl"
local listeners = {}

function i18n.get_locale()
  return current_locale
end

function i18n.set_locale(locale)
  if locale ~= "pl" and locale ~= "en" then return end -- unknown codes are ignored
  current_locale = locale
  for _, listener in ipairs(listeners) do listener(locale) end
end

function i18n.on_locale_change(listener)
  table.insert(listeners, listener)
end

function i18n.t(key)
  local table_for_locale = STRINGS[current_locale] or STRINGS.pl
  return table_for_locale[key] or STRINGS.pl[key] or key
end

-- Test-only accessor (Tests/Lua/spec/i18n_spec.lua) — not called by any
-- gameplay code.
function i18n._key_set(locale)
  local keys = {}
  for k in pairs(STRINGS[locale] or {}) do keys[k] = true end
  return keys
end

return i18n

File: app13_ruby_FastApi/frontend/src/i18n.js
// D-05: a small hand-written i18n store, not a library (PLAN.md §4) — the
// surface area (a few dozen UI strings, two locales) doesn't justify
// i18next/similar. Threat/card CONTENT i18n (description_en/description_pl)
// is a separate concern, handled by `localizedDescription` calls against
// API response fields directly, not by this string table.

const STRINGS = {
  pl: {
    "nav.threats": "Zagrożenia",
    "nav.search": "Szukaj",
    "nav.matrix": "Macierz",
    "nav.about": "O aplikacji",
    "home.title": "Frameworki Bezpieczeństwa",
    "threats.title": "Zagrożenia",
    "threats.filter.severity": "Severity",
    "threats.filter.all": "Wszystkie",
    "threats.empty": "Brak wyników dla podanych filtrów.",
    "detail.overview": "Przegląd",
    "detail.attackVectors": "Wektory Ataku",
    "detail.attackSurface": "Powierzchnia Ataku",
    "detail.mitigations": "Mitigacje",
    "detail.mitigations.empty": "Brak jeszcze zdefiniowanych mitigacji dla tego zagrożenia.",
    "detail.crossReferences": "Powiązania Cross-Framework",
    "detail.crossReferences.empty": "Brak jeszcze zdefiniowanych powiązań dla tego zagrożenia.",
    "codeSample.empty": "Brak jeszcze próbek kodu dla tej mitigacji.",
    "codeSample.attackDemo.label": "ATTACK DEMO — kod podatny, nie używać w produkcji",
    "codeSample.attackDemo.reveal": "Pokaż kod (VULNERABLE)",
    "codeSample.attackDemo.confirmTitle": "Ten kod celowo demonstruje podatność.",
    "codeSample.attackDemo.confirmBody": "Potwierdź, aby go zobaczyć.",
    "codeSample.attackDemo.confirm": "Rozumiem",
    "codeSample.attackDemo.cancel": "Anuluj",
    "search.placeholder": "Szukaj zagrożeń i kart...",
    "search.empty": "Brak wyników.",
    "about.title": "O aplikacji",
    "about.frameworkFree": "Ten frontend jest celowo napisany bez frameworka JS (PLAN.md D-08) — nie jest to skrót, tylko świadoma decyzja projektowa dla porównania kursu."
  },
  en: {
    "nav.threats": "Threats",
    "nav.search": "Search",
    "nav.matrix": "Matrix",
    "nav.about": "About",
    "home.title": "Security Frameworks",
    "threats.title": "Threats",
    "threats.filter.severity": "Severity",
    "threats.filter.all": "All",
    "threats.empty": "No results for the current filters.",
    "detail.overview": "Overview",
    "detail.attackVectors": "Attack Vectors",
    "detail.attackSurface": "Attack Surface",
    "detail.mitigations": "Mitigations",
    "detail.mitigations.empty": "No mitigations defined for this threat yet.",
    "detail.crossReferences": "Cross-Framework References",
    "detail.crossReferences.empty": "No cross-references defined for this threat yet.",
    "codeSample.empty": "No code samples for this mitigation yet.",
    "codeSample.attackDemo.label": "ATTACK DEMO — vulnerable code, do not use in production",
    "codeSample.attackDemo.reveal": "Show code (VULNERABLE)",
    "codeSample.attackDemo.confirmTitle": "This code deliberately demonstrates a vulnerability.",
    "codeSample.attackDemo.confirmBody": "Confirm to view it.",
    "codeSample.attackDemo.confirm": "I understand",
    "codeSample.attackDemo.cancel": "Cancel",
    "search.placeholder": "Search threats and cards...",
    "search.empty": "No results.",
    "about.title": "About",
    "about.frameworkFree": "This frontend is deliberately written with no JS framework (PLAN.md D-08) — not a shortcut, a stated design decision for the course comparison."
  }
};

const STORAGE_KEY = "rubyguard.locale";
const listeners = new Set();

function readStoredLocale() {
  try {
    const stored = localStorage.getItem(STORAGE_KEY);
    return stored === "en" ? "en" : "pl";
  } catch {
    return "pl"; // SR-13.1-equivalent: never crash on a blocked/unavailable localStorage
  }
}

let currentLocale = readStoredLocale();

export function getLocale() {
  return currentLocale;
}

export function setLocale(locale) {
  if (locale !== "pl" && locale !== "en") return; // SR-13.1-equivalent: unknown codes are ignored
  currentLocale = locale;
  try {
    localStorage.setItem(STORAGE_KEY, locale);
  } catch {
    /* localStorage unavailable — locale still applies for this page view */
  }
  listeners.forEach((listener) => listener(locale));
}

export function onLocaleChange(listener) {
  listeners.add(listener);
  return () => listeners.delete(listener);
}

export function t(key) {
  return STRINGS[currentLocale]?.[key] ?? STRINGS.pl[key] ?? key;
}

// Applies `data-i18n="key"` bindings found anywhere under `root` — the
// static-HTML nav links in index.html use this, not JS-rendered views
// (which call `t()` directly inside their own render functions).
export function applyStaticBindings(root = document) {
  root.querySelectorAll("[data-i18n]").forEach((el) => {
    el.textContent = t(el.getAttribute("data-i18n"));
  });
}
